STAN-288: Create Associates microservice for StoryBlocks

moisesnarvaez opened the pull request:
# READY FOR REVIEW

# Summary

Steps followed to complete the requirement requested:

- Get familiar with the code that fetches associates: https://github.com/SU-SWS/saa_victoria_subtheme/blob/2.x/src/associates/src/components/App.jsx
- Create an Associates React component in [GitHub - SU-SWS/saa_alumni: SAA Alumni Homesite](https://github.com/SU-SWS/saa_alumni)
- Create the Associates Directory page in the Stanford StoryBlocks account
- Connect new page and component

# Review By (Date)
- When possible

# Review Tasks

## Setup tasks and/or behavior to test

1. Navigate to https://affiliates--stanford-alumni.netlify.app/stanford-associates/directory/
2. Verify the functionality:
  2.1. List associates in alphabetical order, showing the years on each one.
  2.2. Tab navigation
  2.3. Search by name
  2.4. Filter only new members
4. Review code


# Associated Issues and/or People
- https://fullstacklabs.atlassian.net/browse/STAN-288
- @William-Misiaszek 

## Steps to be Done in the SotyBlocks Prod Account

### Blocks
- Create a new Block Group: "Associates"
- Create new Blocks inside:
  - `associatesDirectory`:
    - Type: Nesteable Block
  - `associatesDirectroryPage`:
    - Type: Content type Block
    - Elements:
      - title (text)
      - intro (rich text)
      - component
        - display name: `associatesDirectory`
        - Allow only: `associatesDirectory`

### Content Page (suggested)

- Create new Content:
  Type: Story
  Name: Directory
  Parent folder: "Stanford Associates"
  Content type: Associates Directory Page
  Prestet: none
- For the new content Story, using the UI editor:
  - In the main content, in the Rick text area, copy the text of https://associatesdesignees.alumni.stanford.edu/

sherakama commented on src/api/contentful/api.js:
```diff
@@ -0,0 +1,23 @@
+    this.accessToken = accessToken;
+  }
+
+  async fethEntries(limit, skip) {
```
```suggestion
  async fetchEntries(limit, skip) {
```

sherakama commented on src/api/contentful/api.js:
Service classes normally go in to `/src/utilities`

sherakama commented on src/api/contentful/associates/index.js:
If you're going with client-side fetching you should move these files to `/src/utilities` otherwise Gatsby is going to try to create an API endpoint for these files. You could add a service layer through the Gatsby API system instead of going directly to ContentFul from the browser but that is a design choice that is up to you.

sherakama commented on src/api/contentful/keys.js:
```diff
@@ -0,0 +1,7 @@
+// It's not necessary to set these as environment variables, they are public and being used on the client side.
```
If you wanted to you could create an API layer to protect all of this using Gatsby's API routing https://www.gatsbyjs.com/docs/reference/functions/routing/

sherakama commented on src/components/page-types/associatesDirectoryPage/Directory/Associate.jsx:
```diff
@@ -0,0 +1,33 @@
+  if (!isEnabled) return null;
+
+  return (
+    <div className="su-flex even:su-bg-black-10">
```
Thanks for using tailwind!

sherakama commented:
@moisesnarvaez I left a few preliminary comments for you

When this is ready for review please tag myself (@sherakama) and (@yvonnetangsu) as reviewers and mark this as ready for review.

moisesnarvaez commented:
Thanks for the early review, I implemented all the suggestions.

The PR is now officially for review.

@sherakama @yvonnetangsu 🙏

sherakama commented:
@moisesnarvaez I think you're missing the masthead (The top bar with logo/nav/search) on the directory page. Can you add it in?

sherakama commented on package.json:
```diff
@@ -23,6 +23,7 @@
     "cnbuilder": "^3.1.0",
     "cookie-parser": "^1.4.6",
     "decanter": "^7.0.0-rc.2",
+    "fast-sort": "^3.4.0",
```
Decent weekly downloads and a recently published version. 👍 to adding this.

sherakama commented on src/api/contentful/associates.js:
```diff
@@ -0,0 +1,6 @@
+import fetchNames from '../../utilities/contentful/associates';
+
+export default async function handler(req, res) {
+  const names = await fetchNames();
+  res.json(names);
```
@moisesnarvaez any error handling required here? What happens when fetchNames fails?

moisesnarvaez replied:
Added a try-catch here

sherakama commented on src/components/page-types/associatesDirectoryPage/Directory/AssociateList.jsx:
```diff
@@ -0,0 +1,52 @@
+      <div>
+        {associates?.map((person, index) => (
+          <Associate
+            // eslint-disable-next-line react/no-array-index-key
```
```suggestion
```
You've added a key now. 👍

moisesnarvaez replied:
I need to add the key because some associates have the same name, and Contentful does not return anything like an ID.

yvonnetangsu commented:
@sherakama As discussed, I'm asking for @rebeccahongsf 's help to review this instead of myself, since she worked on the original widget and is more familiar with the functionality 😄 Thank you @rebeccahongsf in advance 😂

rebeccahongsf commented on src/utilities/contentful/keys.js:
```diff
@@ -0,0 +1,6 @@
+const keys = {
+  space: '0f39zonxf59w',
+  accessToken: '10OGNlSRGeKn81WAaTUxMjVL0nhXFEUszwRJIY7vPeI',
```
@sherakama should we take this opportunity to add the keys to Vault?

moisesnarvaez replied:
I could do it, but I'm not sure if I'm authorized to set the Vault key value.

If yes, can you point me to some Pull Request or file I can use as a guide ?

sherakama replied:
Thanks for moving the fetch to contentful into the API service layer. This will allow us to hide the keys and safely + securely handle the transactions in the future.

sherakama commented on src/components/page-types/associatesDirectoryPage/Directory/Results.jsx:
```diff
@@ -0,0 +1,33 @@
+      <ul>
+        {filteredList?.map((person, index) => (
+          <Associate
+            // eslint-disable-next-line react/no-array-index-key
```
```suggestion
```

moisesnarvaez replied:
I need to add the key because some associates have the same name, and Contentful does not return anything like an ID.

sherakama commented on src/utilities/contentful/keys.js:
```diff
@@ -0,0 +1,6 @@
+const keys = {
+  space: '0f39zonxf59w',
+  accessToken: '10OGNlSRGeKn81WAaTUxMjVL0nhXFEUszwRJIY7vPeI',
```
Thanks @rebeccahongsf & @moisesnarvaez 

I've gone ahead and have put the items into vault. You can pull them down by running `npm run vault` on your local and they will be available on all future builds.  

```suggestion
  space: process.env.ASSOCIATES_CONTENTFUL_SPACE,
  accessToken: process.env.ASSOCIATES_CONTENTFUL_TOKEN,
```

moisesnarvaez replied:
Thank you! I'll implement this

sherakama commented on src/components/page-types/associatesDirectoryPage/legacyDirectory/LegacyDirectory.jsx:
@moisesnarvaez do we need to keep the legacy stuff around? Can we remove it?

moisesnarvaez replied:
Yes, we can remove it

sherakama commented on src/utilities/contentful/api.js:
```diff
@@ -0,0 +1,24 @@
+  async fetchEntries(limit, skip) {
+    const apiUrl = `${this.host}/spaces/${this.space}/entries?access_token=${this.accessToken}&limit=${limit}&skip=${skip}&order=sys.id`;
+
+    const response = await axios.get(apiUrl);
```
Do we need some error handling here?

moisesnarvaez replied:
yes, try-catch added

sherakama commented:
@moisesnarvaez sorry for the slow review but I finally got through it. This is looking and working great. I added a few comments to this PR. Overall, just some cleanup and perhaps a little error handling. Thank you for creating the API layer for fetching contentful data. 

1. Can we remove the Legacy Code?
2. Do you want to add a little error handling for when the contentful fetches fail?
3. I've added the new environment variables to vault. There is a GH suggestion for you to approve in a comment. 
4. Add the Masthead to the page (in Storyblok)

moisesnarvaez commented:
@sherakama thanks for your review! I implemented all the **1-3** requested changes.

About the **4.** Can you please guide me? I added the component based on the other page, what other change is needed?
![image](https://github.com/SU-SWS/saa_alumni/assets/1587535/76032edb-a6c7-4c17-9e80-659b10a26efc)

sherakama commented:
@moisesnarvaez Thanks for the updates.

You were very close on the masthead. Your field name had a capital `M` where the code was expecting the lowercase. I have fixed this in Storyblok for you.

@rebeccahongsf Please review the UX and aesthetics.

yvonnetangsu commented:
@moisesnarvaez I found a small bug (?) that's pretty strange. If I tab over to eg, "C-D" and hit the `spacebar`, it takes me half way down the list of names instead of staying at the top. Strangely if I hit the `enter` key, it works as expected - it's only the `spacebar` that's problematic 🤔 

I'm on Mac Ventura Chrome 116.

cc @sherakama @rebeccahongsf

moisesnarvaez commented:
@yvonnetangsu can you please elaborate what is the expected behaviour?

On other pages (like this Pull Request page for example), it is the same:
- `enter` acts like a click
- `space` acts like a scroll to the next page

rebeccahongsf commented:
That's a great catch @yvonnetangsu ! I was able to reproduce the issue on Google Chrome as well. 

The keyboard accessibility differs from the [prod Associates site](
https://associatesdesignees.alumni.stanford.edu/). On production, the user is able to switch between tabs with the arrow keys. 🤔 

@moisesnarvaez would you be able to update the keyboard accessibility interaction to match production?

moisesnarvaez commented:
@rebeccahongsf I see, thanks for clarifying, working on it

rebeccahongsf commented on src/components/page-types/associatesDirectoryPage/associatesDirectoryPage.js:
```diff
@@ -0,0 +1,46 @@
+          )}
+        </Container>
+
+        <Container className="basic-page su-relative su-flex-grow su-w-full">
```
Also could this be moved this into the `<Container as="main" ...>` component? 😄

rebeccahongsf commented on src/components/page-types/associatesDirectoryPage/associatesDirectoryPage.js:
```diff
@@ -0,0 +1,46 @@
+        <Container as="main" id="main-content">
+          <header className="su-basefont-23">
+            <Container>
+              <Heading align="center" font="serif" id="page-title">
+                {title}
+              </Heading>
```
```suggestion
              <Heading level={1} align="center" font="serif" id="page-title" className="su-max-w-900 su-mb-0 su-rs-py-5 xl:su-rs-py-7 su-type-6 su-mx-auto su-max-w-1200">
                {title}
              </Heading>
```
Let's update this to be an H1 and match the heading styles to other Alumni pages on the site. This will also take care of the top spacing!

rebeccahongsf reviewed (commented):
Thank you @moisesnarvaez for your work on this! I had a few requests regarding the user experience:

- [x] Update Heading to h1 and add styles to match other h1s across the site

- [x] Add spacing to the top and bottom of the main content

- [x] Move the Directory component into the main container

- [ ] Adjust tab styling for mobile

- [ ] Would you be able to add the `Back to top` link to each section so that users can be brought back to the top of the widget? You can view it on [production](https://associatesdesignees.alumni.stanford.edu/) by scrolling towards the bottom of section A (or end of the page)
![image](https://github.com/SU-SWS/saa_alumni/assets/34019925/77de7cee-00fe-47af-a276-73973d6d96dc)

Other than those items, everything else is looking good to me. I'll do another pass once these changes have been implemented. Please do not hesitate to tag me if additional information is needed or if you have any questions. Thanks again! 😃

moisesnarvaez commented:
@rebeccahongsf I implemented all of the requested changes, along with other minor refactoring and CSS adjustments.

rebeccahongsf reviewed (commented):
Thank you for the super quick changes, @moisesnarvaez! 🥳 
A few more design and accessibility tweaks and I think we should be all set!

- [x] Update Container wrapping the Directory component to use `width="full"`

- [x] Adjust Associates wrapper from `<div>` to `<li>` for accessibility

- [ ] Adjust Associates List wrapper from `<div>` to `<ul>` for accessibility

- [ ] Add styles `su-p-0 su-list-none` to Results list to remove additional spacing and possible list styles

moisesnarvaez commented:
@rebeccahongsf Thanks for your kind reviews =). I just pushed the adjustments.

Once this PR is merged, who can take care of publishing this to prod?

rebeccahongsf reviewed (approved):
Thanks again @moisesnarvaez for implementing the requested changes. The associates page looks great to me! 🎉 

As for your question:
>One question, once this PR is merged, who will take care of the required Storyblock changes in Prod?

Either @sherakama or I could move the storyblok blocks to production and setup the page. I defer to Shea for final confirmation though 😄

sherakama commented:
@moisesnarvaez I'll add you to the production instance if you can confirm that the changes are all new items and we aren't having to change or migrate anything existing. Once you've completed the changes in storyblok we will merge the code.

moisesnarvaez commented:
That's correct @sherakama, the items are new.

I have documented them in the description of this PR, in the section called:
**"Steps to be Done in the SotyBlocks Prod Account"**

sherakama commented:
@moisesnarvaez 

I have added you to two new Storyblok spaces for the Prod and Dev instances. Please make the appropriate changes to each Storyblok environment. We will first roll the code through dev to validate deployment went well, and then we will push the code to production. Let me know when you have completed the changes.

moisesnarvaez commented:
@sherakama thanks, all configuration has been replicated.

Dev: https://app.storyblok.com/#/me/spaces/118608/stories/0/0/378510601
Prod: https://app.storyblok.com/#/me/spaces/103430/stories/0/0/378511153

PD: I'm assuming the parent folder for the Content page is "Stanford Associates".

sherakama commented:
Great. Let's start with Dev.

sherakama commented:
Dev looks good in the Storyblok editor. I'll publish it and validate tomorrow that the build works and then we can schedule a release.

sherakama commented:
@moisesnarvaez & @William-Misiaszek Is there a desired release date for production on this functionality?

William-Misiaszek commented:
@sherakama Ideally we release it October 11th in Production (but not published in StoryBlok).


Cc: @moisesnarvaez

sherakama commented:
Ok thanks. I'll put a release event on the calendar for the 11th.
